tools/clang/unittests/HLSLExec/LongVectors.cpp

```diff
@@ -656,200 +510,157 @@ void runAndVerify(const TestConfig &Config, OP_TYPE OpType,
+//
+// Most Ops have a 1:1 mapping of input to output, and so can use the generic
+// ExpectedBuilder.
+//
+// Ops that differ from this pattern can specialized ExpectedBuilder as
+// necessary.
+//
+
+// Op - specializations are expected to have a ValidationConfig member and an
+// appropriate overloaded function call operator.
+template <OpType OP, typename T> struct Op;
 
-  InputSets<T, 1> Inputs = buildTestInputs<T, 1>(Config, VectorSize);
+// ExpectedBuilder - specializations are expected to have buildExpectedData
+// member functions.
+template <OpType OP, typename T> struct ExpectedBuilder;
 
-  std::vector<OUT_TYPE> Expected;
-  Expected.reserve(Inputs[0].size());
+// Default Validation configuration - for non-floating point types uses ULP.
+// For other types requires exact matches.
+template <typename T> struct DefaultValidation {
+  ValidationConfig ValidationConfig;
 
-  for (size_t I = 0; I < Inputs[0].size(); ++I)
-    Expected.push_back(Calc(Inputs[0][I]));
+  DefaultValidation() {
+    if constexpr (isFloatingPointType<T>())
+      ValidationConfig = ValidationConfig::Ulp(1.0f);
+  }
+};
 
-  runAndVerify(Config, OpType, Inputs, Expected, ExtraDefines,
-               ValidationConfig);
-}
+// Strict Validation - for all types require exact matches.
+struct StrictValidation {
+  ValidationConfig ValidationConfig;
+};
 
-template <typename T, typename OUT_TYPE, typename OP_TYPE>
-void dispatchBinaryTest(const TestConfig &Config,
-                        const ValidationConfig &ValidationConfig,
-                        OP_TYPE OpType, size_t VectorSize,
-                        OUT_TYPE (*Calc)(T, T), std::string ExtraDefines = "") {
-  InputSets<T, 2> Inputs = buildTestInputs<T, 2>(Config, VectorSize);
+// Macros to build up common patterns of Op definitions
 
-  std::vector<OUT_TYPE> Expected;
-  Expected.reserve(Inputs[0].size());
+#define OP_1(OP, VALIDATION, IMPL)                                             \
+  template <typename T> struct Op<OP, T> : VALIDATION {                        \
+    T operator()(T A) { return IMPL; }                                         \
+  }
 
-  for (size_t I = 0; I < Inputs[0].size(); ++I) {
-    size_t Index1 = (Config.ScalarInputFlags & (1 << 1)) ? 0 : I;
-    Expected.push_back(Calc(Inputs[0][I], Inputs[1][Index1]));
+#define OP_2(OP, VALIDATION, IMPL)                                             \
+  template <typename T> struct Op<OP, T> : VALIDATION {                        \
+    T operator()(T A, T B) { return IMPL; }                                    \
   }
 
-  runAndVerify(Config, OpType, Inputs, Expected, ExtraDefines,
-               ValidationConfig);
-}
+#define OP_3(OP, VALIDATION, IMPL)                                             \
+  template <typename T> struct Op<OP, T> : VALIDATION {                        \
+    T operator()(T A, T B, T C) { return IMPL; }                               \
+  }
+
+#define DEFAULT_OP_1(OP, IMPL) OP_1(OP, DefaultValidation<T>, IMPL)
```

Comment:
nit: I'm wondering if 'UNARY_OP' would be a better name here. I feel like the default can be implied.

Reply:
The `_1`, `_2` and `_3` suffixes specify ARITY.  The "DEFAULT" part refers to the validation type.  We already have `OP_1` etc. for ones where the validation type needs to be explicitly specified.